### ISSUE TITLE
Some feedback on `contribute_releases`

### ISSUE THREAD

**Issue by llrs** [r-releases/releases#2]
I think {releases} should delegate the work of registering new things to the "official" package: universe: https://github.com/ropensci/universe/issues/1 . 
It should focus on handling what is r-releases control the formatting of the PR and where to send it (specially considering https://github.com/r-releases/help/issues/10).

Apparently it needs to pull all the code from the r-releases/contributions without considering that the user might already have it. 
In the process is too chatty and might find some problems (see the first attempt):

```
contribute_releases("https://github.com/llrs/experDesign")
Getting GitHub identity.
Checking if fork 'https://github.com/llrs/r-releases' already exists.
Cloning fork 'https://github.com/llrs/r-releases'
Transferred 329 of 329 objects...done!
Checked out 83 of 83 commits... done!
Preparing pull request.
Transferred 573 of 573 objects...done!
[new]     85f03366717a348a17a1 refs/remotes/upstream/glmnet
[new]     377c9f587618bac34be4 refs/remotes/upstream/main
[new]     fa8013c9300e8a362eb2 refs/remotes/upstream/mmrm
[new]     41bac62fbd92c0196df2 refs/remotes/upstream/test-no-release
Performing fast-forward merge, no commit needed
Error in gert::git_commit(message = title, repo = repo) : 
  No staged files to commit. Run git_add() to select files.
> contribute_releases(experDesign = "https://github.com/llrs/experDesign")
Getting GitHub identity.
Checking if fork 'https://github.com/llrs/r-releases' already exists.
Cloning fork 'https://github.com/llrs/r-releases'
Transferred 329 of 329 objects...done!
Checked out 83 of 83 commits... done!
Preparing pull request.
Transferred 573 of 573 objects...done!
[new]     85f03366717a348a17a1 refs/remotes/upstream/glmnet
[new]     377c9f587618bac34be4 refs/remotes/upstream/main
[new]     fa8013c9300e8a362eb2 refs/remotes/upstream/mmrm
[new]     41bac62fbd92c0196df2 refs/remotes/upstream/test-no-release
Performing fast-forward merge, no commit needed
Looking up https credentials for https://github.com/llrs/r-releases
Transferred 4 of 4 objects...done!
[status] refs/heads/branch_5b82a77f7f289: unchanged
[new]     2e11075f8022ffe50f20 refs/remotes/origin/branch_5b82a77f7f289
Submitting pull request.
Pull request successfully posted to 'https://github.com/r-releases/contributions/pull/146'. Please give the bot time to review your contribution.
> contribute_releases(experDesign = "https://github.com/llrs/experDesign")
Getting GitHub identity.
Checking if fork 'https://github.com/llrs/r-releases' already exists.
Cloning fork 'https://github.com/llrs/r-releases'
Transferred 898 of 898 objects...done!
Checked out 83 of 83 commits... done!
Preparing pull request.
Transferred 8 of 8 objects...done!
[new]     85f03366717a348a17a1 refs/remotes/upstream/glmnet
[new]     377c9f587618bac34be4 refs/remotes/upstream/main
[new]     fa8013c9300e8a362eb2 refs/remotes/upstream/mmrm
[new]     41bac62fbd92c0196df2 refs/remotes/upstream/test-no-release
Performing fast-forward merge, no commit needed
Looking up https credentials for https://github.com/llrs/r-releases
Transferred 1 of 1 objects...done!
[status] refs/heads/branch_5b82a3781b47f: unchanged
[new]     5d45603039d48b6d6482 refs/remotes/origin/branch_5b82a3781b47f
Submitting pull request.
Pull request successfully posted to 'https://github.com/r-releases/contributions/pull/147'. Please give the bot time to review your contribution.
```

In addition there is no check if the package is already on r-releases (See [#146](https://github.com/r-releases/contributions/pull/146), first attempt) and  ([#147](https://github.com/r-releases/contributions/pull/147) second attempt).
This could be done by checking open issues and providing an informative message to users.

BTW, in my opinion, ideally each branch created would have either the name of the repo/package to add or ideally one meaningful branch (YYYYMMDDHHMMSS?) if there are multiple repositories.

**Comment by llrs** [comment]
In addition I added some checks to the function in a recent commit https://github.com/r-releases/releases/commit/25ddd2e165626e219dd0cacf5127707727c62ded. Sorry I realized to late to create a PR instead of committing directly.

**Comment by wlandau** [comment]
> I think {releases} should delegate the work of registering new things to the "official" package: universe: https://github.com/ropensci/universe/issues/1 .

Would you elaborate? I am not sure what you mean.

> It should focus on handling what is r-releases control the formatting of the PR and where to send it (specially considering https://github.com/r-releases/help/issues/10).

Do you mean that `contribute_releases()` should run `r.releases.internals::assert_package()` to check the contribution before the PR is opened?

> Apparently it needs to pull all the code from the r-releases/contributions without considering that the user might already have it.

The fork is only created on GitHub if the user does not have it. The clone is always in a `tempfile()`, so it's new every time.

> In the process is too chatty and might find some problems (see the first attempt):

Unfortunately we are relying a lot of the `gert` package, which is already chatty. Maybe that can be simplified, but I think it's good to err on the side of verbosity so users understand where an error comes from and can post a more informative bug report.

> In addition I added some checks to the function in a recent commit https://github.com/r-releases/releases/commit/25ddd2e165626e219dd0cacf5127707727c62ded. Sorry I realized to late to create a PR instead of committing directly.

Yes, let's please use PRs. I just added branch protection in Settings.

**Comment by llrs** [comment]
> > I think {releases} should delegate the work of registering new things to the "official" package: universe: [ropensci/universe#1](https://github.com/ropensci/universe/issues/1) .
> 
> Would you elaborate? I am not sure what you mean.

Yes, sorry, I didn't start the "issue" the best way, as this is something beyond the `contribute_releases`.
Updating, modifying and adding a package to a r-universe is (or might become) a common process. 
Maybe the way to register a package in r-releases (and the other repo r-contributor?) is specific to this project. 
But CRANhaven is doing a similar process and in the future there might be other repositories, so I think there could be more coordination between projects via a package that might help with some common patterns. 
In my opinion {r.releases.internals} and if it is possible {releases} should try to use these methods. 

> > It should focus on handling what is r-releases control the formatting of the PR and where to send it (specially considering [r-releases/help#10](https://github.com/r-releases/help/issues/10)).
> 
> Do you mean that `contribute_releases()` should run `r.releases.internals::assert_package()` to check the contribution before the PR is opened?

Yes, and `r.releases.internals::assert_release_exists` too to help contributors and provide feedback to the developer/user. 

> 
> > Apparently it needs to pull all the code from the r-releases/contributions without considering that the user might already have it.
> 
> The fork is only created on GitHub if the user does not have it. The clone is always in a `tempfile()`, so it's new every time.

Exactly! If I contribute multiple times in the same R session I need a new clone each time instead of updating the temp folder created. 
In addition, if the user has already cloned the repository there is no way to use that and avoid the download and update of the whole history of the repository.

> 
> > In the process is too chatty and might find some problems (see the first attempt):
> 
> We are relying a lot of the `gert` package, which is already very chatty, and there are a lot of Git and GitHub steps in this workflow. Maybe the messages can be simplified, but that would require a lot of manual work. I think it's good to err on the side of verbosity so users understand where an error comes from and can post a more informative bug report.

I agree that useful error messages are great. 
But we should process them to make it easier for us when users will complain it didn't work. 
For instance, I didn't understand the verbosity help with the error in my first attempt and I still don't know went wrong. 

> 
> > In addition I added some checks to the function in a recent commit [25ddd2e](https://github.com/r-releases/releases/commit/25ddd2e165626e219dd0cacf5127707727c62ded). Sorry I realized to late to create a PR instead of committing directly.
> 
> Yes, let's please use PRs. I just added branch protection in Settings.

Thanks, I thought I was already in a branch :sweat: This will help

**Comment by wlandau** [comment]
> Maybe the way to register a package in r-releases (and the other repo r-contributor?) is specific to this project.

Yes, the process of contribution is very specific to R-releases. Of course, others can borrow the infrastructure, but R-releases does not aim to solve the more general problem of maintaining a collaborative R-universe. That is beyond scope.

> Exactly! If I contribute multiple times in the same R session I need a new clone each time instead of updating the temp folder created.
> In addition, if the user has already cloned the repository there is no way to use that and avoid the download and update of the whole history of the repository.

We could maintain a persistent copy somewhere in `tools::R_user_dir()`. And to prevent merge conflicts, it should probably clone/pull from the upstream remote at https://github.com/r-releases/contributions before pushing a new branch to the fork.

> But we should process them to make it easier for us when users will complain it didn't work.

Rather than remove messages, what if we group them? We could clump them in sections under informative headers (i.e. `cli::cli_h1()`)

**Comment by wlandau** [comment]
> Yes, and r.releases.internals::assert_release_exists too to help contributors and provide feedback to the developer/user.

assert_package() calls assert_release_exists().

**Comment by wlandau** [comment]
Just found out about the `verbose` argument in `gert` functions. This will help cut down on verbosity.
